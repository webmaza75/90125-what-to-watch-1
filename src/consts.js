export const ALL_GENRES = `All genres`;
export const BASE_URL = `https://es31-server.appspot.com`;
export const ValidationErrors = {
  INVALID_EMAIL: `child \"email\" fails because [\"email\" is not allowed to be empty]`,
  INVALID_PASSWORD: `child \"password\" fails because [\"password\" is not allowed to be empty]`,
  INVALID_EMAIL_AND_PASSWORD: `\"email\" and \"password\" are not allowed to be empty`,
};
export const RatingLevels = {
  AWESOME: `Awesome`,
  VERY_GOOD: `Very good`,
  GOOD: `Good`,
  NORMAL: `Normal`,
  BAD: `Bad`
};
export const MAX_GENRES = 10;
export const MAX_FILMS = 20;
export const tabs = [
  `Overview`,
  `Details`,
  `Reviews`
];
export const Tabs = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
export const NO_COMMENTS = `No comments`;
