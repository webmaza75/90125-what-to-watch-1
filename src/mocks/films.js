const filmList = [
  {
    backgroundColor: `#9B7E61`,
    backgroundImage: `https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    director: `Steven Spielberg`,
    genre: `Adventure`,
    id: 1,
    isFavorite: false,
    title: `War of the Worlds`,
    posterImage: `https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg`,
    picture: `https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg`,
    src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    rating: 6.5,
    ratingLevel: `Good`,
    released: 2005,
    runTime: 116,
    scoresCount: 386834,
    starring: [
      `Tom Cruise`,
      `Dakota Fanning`,
      `Tim Robbins`,
    ],
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`
  },
  {
    backgroundColor: `#F1E9CE`,
    backgroundImage: `https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg`,
    description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
    director: `Justin Kurzel`,
    genre: `Drama`,
    id: 2,
    isFavorite: false,
    title: `Macbeth`,
    posterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
    picture: `https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg`,
    src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 6.6,
    ratingLevel: `Good`,
    released: 2015,
    runTime: 113,
    scoresCount: 48798,
    starring: [
      `Michael Fassbender`,
      `Marion Cotillard`,
      `Jack Madigan`
    ],
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`
  },
  {
    backgroundColor: `#D8E3E5`,
    backgroundImage: `https://es31-server.appspot.com/wtw/static/film/background/Moonrise_Kingdom.jpg`,
    description: `A pair of young lovers flee their New England town, which causes a local search party to fan out to find them.`,
    director: `Wes Anderson`,
    genre: `Adventure`,
    id: 3,
    isFavorite: false,
    title: `Moonrise Kingdom`,
    posterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Moonrise_Kingdom.jpg`,
    picture: `https://es31-server.appspot.com/wtw/static/film/preview/moonrise-kingdom.jpg`,
    src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 7.8,
    ratingLevel: `Good`,
    released: 2012,
    runTime: 94,
    scoresCount: 291183,
    starring: [
      `Jared Gilman`,
      `Kara Hayward`,
      `Bruce Willis`
    ],
    videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`
  }
];

export default filmList;
