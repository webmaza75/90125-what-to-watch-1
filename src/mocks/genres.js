const genres = [
  `Comedy`,
  `Crime`,
  `Documentary`,
  `Drama`,
  `Horror`,
  `Kids & Family`,
  `Romance`,
  `Sci-Fi`,
  `Thriller`
];

export default genres;
