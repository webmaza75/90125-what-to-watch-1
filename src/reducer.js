import ActionTypes from './action-types.js';
import filmList from './mocks/films.js';
import {ALL_GENRES} from './consts.js';

const initialState = {
  films: filmList,
  filter: ALL_GENRES,
  filmsGroup: filmList
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.CHANGE_GENRE:
      return {
        ...state,
        filter: action.payload
      };
    case ActionTypes.GET_FILMS_BY_GENRE:
      if (action.payload.filter === ALL_GENRES) {
        return initialState;
      }
      return {
        ...state,
        filmsGroup: action.payload.films.filter(({genre}) => genre.includes(action.payload.filter))
      };
  }
  return state;
};

export default reducer;
