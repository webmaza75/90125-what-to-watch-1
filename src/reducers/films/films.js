import ActionTypes from '../../actions/action-types.js';
import {ALL_GENRES} from '../../consts.js';
import {ActionCreator} from '../../actions/actions.js';
import {sortComments} from '../../utils.js';
import {getMovieRatingLevel} from '../../utils.js';

const initialState = {
  films: [],
  filter: ALL_GENRES,
  comments: []
};

export const transform = (data) => ({
  backgroundColor: data.background_color,
  backgroundImage: data.background_image,
  description: data.description,
  director: data.director,
  genre: data.genre,
  id: data.id,
  isFavorite: data.is_favorite,
  title: data.name,
  posterImage: data.poster_image,
  picture: data.preview_image,
  src: data.preview_video_link,
  rating: data.rating,
  ratingLevel: getMovieRatingLevel(data.rating),
  released: data.released,
  runTime: data.run_time,
  scoresCount: data.scores_count,
  starring: data.starring,
  videoLink: data.video_link
});

const Operation = {
  loadFilms: () => (dispatch, getState, api) => {
    return api.get(`/films`)
    .then((response) => {
      const result = response.data.map(transform);
      dispatch(ActionCreator.loadFilms(result));
    });
  },
  loadComments: (filmId) => (dispatch, getState, api) => {
    return api.get(`/comments/${filmId}`)
    .then((response) => {
      const result = response.data.sort(sortComments);
      dispatch(ActionCreator.loadComments(result));
    });
  },
  addComment: (id, params) => (dispatch, getState, api) => {
    return api
      .post(`/comments/${id}`, params)
      .then((res) => {
        dispatch(ActionCreator.addComment(res.data));
      }).catch((error) => {
        dispatch(ActionCreator.submitCommentsError(error.response.data.error));
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.LOAD_FILMS:
      return ({
        ...state,
        films: action.payload,
      });
    case ActionTypes.CHANGE_GENRE:
      return {
        ...state,
        filter: action.payload
      };
    case ActionTypes.LOAD_COMMENTS:
      return ({
        ...state,
        comments: action.payload,
      });
    case ActionTypes.ADD_COMMENT:
      return ({
        ...state,
        comments: action.payload,
      });
  }
  return state;
};

export {
  Operation,
  reducer
};
