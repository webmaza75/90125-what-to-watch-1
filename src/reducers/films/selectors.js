import {createSelector} from 'reselect';
import NameSpace from '../name-spaces.js';
import {ALL_GENRES} from '../../consts.js';

const NAME_SPACE = NameSpace.FILMS;

export const getFilms = (state) => {
  return state[NAME_SPACE].films;
};

export const getActiveFilter = (state) => {
  return state[NAME_SPACE].filter;
};

export const getMaxShowFilms = (state) => {
  return state[NAME_SPACE].maxShowFilms;
};

export const getFilmsByGenre = createSelector(
    getFilms,
    getActiveFilter,
    getMaxShowFilms,
    (films, filter, maxShowFilms) => {
      if (filter === ALL_GENRES) {
        return films.slice(0, maxShowFilms);
      }
      return films.filter((item) => item.genre === filter).slice(0, maxShowFilms);
    }
);

export const getGenres = createSelector(
    getFilms,
    (films) => {
      const allGenres = films.map(({genre}) => genre);
      return [ALL_GENRES, ...new Set(allGenres)];
    }
);

export const getFavoriteList = (state) => {
  return state[NAME_SPACE].favorites;
};

export const getMovieById = (state, filmId) => {
  return state[NAME_SPACE].films.find((item) => item.id === +filmId);
};

export const getComments = (state) => {
  return state[NAME_SPACE].comments;
};

export const getPromo = (state) => {
  return state[NAME_SPACE].promo;
};
