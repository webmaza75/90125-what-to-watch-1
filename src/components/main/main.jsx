import React, {Fragment, PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import FilmList from '../film-list/film-list.jsx';
import FilmItem from '../film-item/film-item.jsx';
import {ActionCreator} from '../../actions/actions.js';
import GenreList from '../genre-list/genre-list.jsx';
import withActiveItem from '../../hocs/with-active-item/with-active-item.js';
import {
  getActiveFilter,
  getFilmsByGenre,
  getGenres,
  getPromo,
  getPlayState
} from '../../reducers/films/selectors.js';
import Footer from '../footer/footer.jsx';
import {Operation} from '../../reducers/films/films.js';
import FilmPromo from '../film-promo/film-promo.jsx';
import {itemShape} from '../../models.js';
import ShowMoreButton from '../show-more-button/show-more-button.jsx';
import Player from '../player/player.jsx';

const FilmListWrapped = withActiveItem(FilmList);

class Main extends PureComponent {
  constructor(props) {
    super(props);

    this._handleMenuClick = this._handleMenuClick.bind(this);
  }

  componentDidMount() {
    const {onLoadPromo} = this.props;
    onLoadPromo();
  }

  componentWillUnmount() {
    const {onResetMaxShowFilms} = this.props;
    onResetMaxShowFilms();
  }

  render() {
    const {
      filmsGroup,
      filter,
      genres,
      promo,
      showPlayer
    } = this.props;

    if (!promo || !promo.id) {
      return null;
    }

    if (showPlayer) {
      return <Player movie={promo} />;
    }

    const {
      title,
      genre,
      released,
      backgroundImage,
      id,
      isFavorite,
      posterImage
    } = promo;

    return <Fragment>
      <section className="movie-card">
        <FilmPromo
          title={title}
          genre={genre}
          released={released}
          backgroundImage={backgroundImage}
          id={id}
          isFavorite={isFavorite}
          posterImage={posterImage}
        />
      </section>

      <div className="page-content">
        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>

          <GenreList
            genres={genres}
            onChange={this._handleMenuClick}
            activeItem={filter}
          />

          <FilmListWrapped
            films={filmsGroup}
          />

          <ShowMoreButton />

        </section>
        <Footer />
      </div>
    </Fragment>;
  }

  _handleMenuClick(genre) {
    const {onChangeFilter} = this.props;
    onChangeFilter(genre);
  }
}

Main.propTypes = {
  filter: PropTypes.string,
  filmsGroup: PropTypes.arrayOf(FilmItem.propTypes.item),
  onChangeFilter: PropTypes.func,
  genres: PropTypes.arrayOf(PropTypes.string),
  promo: itemShape,
  onLoadPromo: PropTypes.func,
  showPlayer: PropTypes.bool,
  onResetMaxShowFilms: PropTypes.func
};

const mapStateToProps = (state) => ({
  filter: getActiveFilter(state),
  filmsGroup: getFilmsByGenre(state),
  genres: getGenres(state),
  promo: getPromo(state),
  showPlayer: getPlayState(state)
});

const mapDispatchToProps = {
  onChangeFilter: ActionCreator.changeFilter,
  onLoadPromo: Operation.loadPromo,
  onToggleFavorite: Operation.toggleFavorite,
  onResetMaxShowFilms: ActionCreator.resetMaxShowFilms
};

export {Main};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Main);
