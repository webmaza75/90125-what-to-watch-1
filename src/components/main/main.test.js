import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom';
import {Provider} from 'react-redux';
import {createStore} from 'redux';

import {Main} from './main.jsx';
import films from '../../mocks/films.js';
import {ALL_GENRES} from '../../consts.js';
import genres from '../../mocks/genres.js';
import reducer from '../../reducers/reducer.js';

window.HTMLMediaElement.prototype.play = () => {};

it(`Main correctly renders after relaunch`, () => {
  const tree = renderer
    .create(<Provider store={createStore(reducer)}>
      <MemoryRouter initialEntries={[`/`]}>
        <Main
          filter={ALL_GENRES}
          filmsGroup={films}
          genres={genres}
          onChangeFilter={jest.fn()}
          onLoadPromo={jest.fn()}
          promo={films[0]}
          onResetMaxShowFilms={jest.fn()}
        />
      </MemoryRouter>
    </Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
