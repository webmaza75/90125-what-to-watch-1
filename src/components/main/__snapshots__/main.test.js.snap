// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main correctly renders after relaunch 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="War of the Worlds"
        src="https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link "
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="War of the Worlds poster"
            height="327"
            src="https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            War of the Worlds
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Adventure
            </span>
            <span
              className="movie-card__year"
            >
              2005
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Comedy
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Crime
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Documentary
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Drama
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Horror
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Kids & Family
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Romance
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Sci-Fi
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Thriller
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            href="/film/1"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <img
                alt="War of the Worlds"
                height="175"
                src="https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg"
                width="280"
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/1"
              onClick={[Function]}
            >
              War of the Worlds
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            href="/film/2"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <img
                alt="Macbeth"
                height="175"
                src="https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg"
                width="280"
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/2"
              onClick={[Function]}
            >
              Macbeth
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            href="/film/3"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <img
                alt="Moonrise Kingdom"
                height="175"
                src="https://es31-server.appspot.com/wtw/static/film/preview/moonrise-kingdom.jpg"
                width="280"
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/3"
              onClick={[Function]}
            >
              Moonrise Kingdom
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
